src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -379,13 +381,22 @@ Gmres<VarsType, Preconditioner, LinearSolverRegistrars>::solve(
   while (not has_converged) {
     const auto& initial_guess = *initial_guess_in_solution_out;
     auto& initial_operand = basis_history_[0];
-    std::apply(
-        linear_operator,
-        std::tuple_cat(std::forward_as_tuple(make_not_null(&initial_operand),
-                                             initial_guess),
-                       operator_args));
-    initial_operand *= -1.;
-    initial_operand += source;
+    // Apply the linear operator to the initial guess. This can be skipped if
+    // the initial guess is zero, because then the linear operator applied to it
+    // is also zero.
+    if (alg::all_of(initial_guess, [](const double element) {
+          return equal_within_roundoff(element, 0.);
```

Comment:
The check for zero data here adds a little overhead to the case where the data is _not_ zero. This overhead doesn't seem to be significant, since the first non-zero datapoint will break the `alg::all_of`. Nevertheless, if anyone has a better idea how to do this, I'd be glad to hear it!